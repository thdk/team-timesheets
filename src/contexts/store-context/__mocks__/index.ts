export const useStore = jest.fn();
