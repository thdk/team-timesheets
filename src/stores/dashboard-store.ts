import { observable, reaction, computed, action } from "mobx";
import { ICollection, Collection, RealtimeMode } from "firestorable";
import type firebase from "firebase";

import moment from 'moment';

import { IRootStore } from "./root-store";
import { IGroupedRegistrations } from "./registration-store/registration-store";
import * as deserializer from '../../common/serialization/deserializer';
import * as serializer from '../../common/serialization/serializer';
import { TimePeriod } from "../components/time-period-select";
import { IRegistration, IRegistrationData } from "../../common/dist";

export interface IDashboardStore {
    readonly setProjectFilter: (projectId: string | undefined) => void;
    readonly setUserFilter: (userId: string | undefined) => void;
    readonly setTaskFilter: (taskId: string | undefined) => void;
    readonly setTimePeriodFilter: (timePeriod: TimePeriod) => void;

    readonly registrationsGroupedByUser: IGroupedRegistrations<string>[];
    readonly registrationsGroupedByTask: IGroupedRegistrations<string>[];
    readonly registrationsGroupedByProject: IGroupedRegistrations<string>[];

    readonly projectFilterValue?: string;
    readonly taskFilterValue?: string;
    readonly userFilterValue?: string;
    readonly timePeriodFilterValue?: TimePeriod;

}

export class DashboardStore implements IDashboardStore {
    private readonly registrationsField: ICollection<IRegistration, IRegistrationData>;

    @observable.ref private userFilterValueField: string | undefined = undefined;
    @observable.ref private taskFilterValueField: string | undefined = undefined;
    @observable.ref private projectFilterValueField: string | undefined = undefined;
    @observable.ref private timePeriodFilterField: TimePeriod | undefined = undefined;

    constructor(
        rootStore: IRootStore,
        {
            firestore,
        }: {
            firestore: firebase.firestore.Firestore,
        },
    ) {
        this.registrationsField = new Collection<IRegistration, IRegistrationData>(
            firestore,
            "registrations",
            {
                realtimeMode: RealtimeMode.on,
                deserialize: deserializer.convertRegistration,
                serialize: serializer.convertRegistration,
                name: "Dashboard registrations",
            },
            {
                // logger: console.log
            },
        );

        const updateRegistrationQuery = () => {
            if (!rootStore.user.divisionUser || !this.timePeriodFilterField) {
                this.registrationsField.query = null;
            } else {
                this.registrationsField.query = ref =>
                    this.userIdFilter(
                        this.projectIdFilter(
                            this.endDateFilter(
                                this.startDateFilter(
                                    ref.where("deleted", "==", false)
                                )
                            )
                        )
                    );
            }
        };

        // update the query of the registration collection each time...
        // -- a filter changes (userId, startDate, endDate)
        // TODO: add a Filter button in UI to group all filter changes into one new search query
        reaction(() => this.timePeriodFilterField, updateRegistrationQuery);
        reaction(() => this.userFilterValue, updateRegistrationQuery);
        reaction(() => this.projectFilterValue, updateRegistrationQuery);
        reaction(() => rootStore.user.divisionUser, updateRegistrationQuery);
    }

    @computed
    private get startDateFilter() {
        const startDate = this.startDate;
        return (query: firebase.firestore.Query) => startDate ? query.where("date", ">=", startDate) : query;
    }

    @computed
    private get endDateFilter() {
        const endDate = this.endDate;
        return (query: firebase.firestore.Query) => endDate ? query.where("date", "<=", endDate) : query;
    }

    @computed
    private get userIdFilter() {
        const userId = this.userFilterValue;
        return (query: firebase.firestore.Query) => userId ? query.where("userId", "==", userId) : query;
    }

    @computed
    private get projectIdFilter() {
        return (query: firebase.firestore.Query) => this.projectFilterValue
            ? query.where("project", "==", this.projectFilterValue)
            : query;
    }

    @computed
    get registrationsGroupedByUser() {
        // no need to group when userId is included in the filter
        if (this.registrations.length === 0) return [];

        const registrationsPerUser = this.createRegistrationsGroup(reg => reg.userId);

        return registrationsPerUser;
    }

    @computed
    get registrationsGroupedByProject() {
        if (this.registrations.length === 0) return [];

        return this.createRegistrationsGroup(reg => reg.project || "")
            .sort((a, b) => b.totalTime - a.totalTime);;
    }

    @computed
    get registrationsGroupedByTask() {
        if (this.registrations.length === 0) return [];

        return this.createRegistrationsGroup(reg => reg.task || "")
            .sort((a, b) => b.totalTime - a.totalTime);
    }

    @computed
    private get startDate() {
        const today = moment(new Date());
        switch (this.timePeriodFilterField) {
            case TimePeriod.ThisWeek:
                return today.clone().startOf('week').toDate();
            case TimePeriod.ThisMonth:
                return today.clone().startOf('month').toDate();
            case TimePeriod.LastMonth:
                return today.clone().subtract(1, 'months').startOf('month').toDate();
            case TimePeriod.ThisYear:
                return today.clone().startOf('year').toDate();
            case TimePeriod.LastYear:
                return today.clone().subtract(1, 'years').startOf('year').toDate();
            default:
                return undefined;
        }
    }

    @computed
    private get endDate() {
        const today = moment(new Date());
        switch (this.timePeriodFilterField) {
            case TimePeriod.ThisWeek:
                return today.clone().endOf('week').toDate();
            case TimePeriod.ThisMonth:
                return today.clone().endOf('month').toDate();
            case TimePeriod.LastMonth:
                return today.clone().subtract(1, 'months').endOf('month').toDate();
            case TimePeriod.ThisYear:
                return today.clone().endOf('year').toDate();
            case TimePeriod.LastYear:
                return today.clone().subtract(1, 'years').endOf('year').toDate();
            default:
                return undefined;
        }
    }

    public get taskFilterValue() {
        return this.taskFilterValueField;
    }

    public get userFilterValue() {
        return this.userFilterValueField;
    }

    public get projectFilterValue() {
        return this.projectFilterValueField;
    }

    public get timePeriodFilterValue() {
        return this.timePeriodFilterField;
    }

    @computed
    private get registrations() {
        return this.registrationsField.docs
            .filter(doc => doc.data!.isPersisted) // don't display drafts (TODO: move to collection.ts?)
    }

    @action
    public setProjectFilter(projectId: string | undefined) {
        this.projectFilterValueField = projectId;
    }

    @action
    public setTaskFilter(taskId: string | undefined) {
        this.taskFilterValueField = taskId;
    }

    @action
    public setUserFilter(userId: string | undefined) {
        this.userFilterValueField = userId;
    }

    @action setTimePeriodFilter(timePeriod: TimePeriod) {
        this.timePeriodFilterField = timePeriod;
    }

    private createRegistrationsGroup = <T>(keySelector: (reg: IRegistration) => T): IGroupedRegistrations<T>[] => {
        const groupedMap = new Map<T, IGroupedRegistrations<T>>();

        this.registrations
            .forEach(c => {
                const key = keySelector(c.data!);
                const group = groupedMap.get(key);
                if (group) {
                    group.totalTime += c.data!.time || 0;
                    group.registrations.push(c);
                }
                else {
                    groupedMap.set(key, {
                        groupKey: keySelector(c.data!),
                        registrations: [c],
                        totalTime: c.data!.time || 0,
                        isCollapsed: true
                    });
                }
            });

        return Array.from(groupedMap.values());
    }
}