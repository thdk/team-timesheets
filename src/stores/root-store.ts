import { IRegistrationsStore, RegistrationStore } from "./registration-store";
import { IConfigStore, ConfigStore } from "./config-store";
import { IUserStore, UserStore } from "./user-store";
import { IViewStore, ViewStore } from "./view-store";
import { RouterStore } from "mobx-router";
import { IReportStore, ReportStore } from "./report-store";
import { DashboardStore, IDashboardStore } from "./dashboard-store";
import { IProjectStore, ProjectStore } from "./project-store";
import { FavoriteStore } from "./favorite-store";

export interface IRootStore {
    readonly user: IUserStore;
    readonly view: IViewStore;
    readonly router: RouterStore;
    readonly timesheets: IRegistrationsStore;
    readonly reports: IReportStore;
    readonly config: IConfigStore;
    readonly projects: IProjectStore;
    readonly favorites: FavoriteStore;
    readonly dashboard: IDashboardStore;
}

export class Store implements IRootStore {
    public readonly timesheets: IRegistrationsStore;
    public readonly view: IViewStore;
    public readonly user: IUserStore;
    public readonly config: IConfigStore;
    public readonly router = new RouterStore();
    public readonly reports: IReportStore;
    public readonly dashboard: IDashboardStore;
    public readonly projects: IProjectStore;
    public readonly favorites: FavoriteStore;

    constructor() {
        this.user = new UserStore(this);
        this.view = new ViewStore(this);
        this.config = new ConfigStore(this);
        this.timesheets = new RegistrationStore(this);
        this.reports = new ReportStore(this);
        this.dashboard = new DashboardStore(this);
        this.projects = new ProjectStore(this);
        this.favorites = new FavoriteStore(this);
    }
};
