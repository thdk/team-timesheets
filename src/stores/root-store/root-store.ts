import { RouterStore } from "mobx-router";

import { IRegistrationsStore, RegistrationStore } from "../registration-store/registration-store";
import { IConfigStore, ConfigStore } from "../config-store";
import { IUserStore, UserStore } from "../user-store";
import { IViewStore, ViewStore } from "../view-store";
import { IReportStore, ReportStore } from "../report-store/report-store";
import { DashboardStore, IDashboardStore } from "../dashboard-store";
import { ProjectStore } from "../project-store";
import { FavoriteStore } from "../favorite-store";
import { DivisionStore  } from "../division-store/division-store";

export interface IRootStore extends Store { };

export class Store implements IRootStore {
    public readonly timesheets: IRegistrationsStore;
    public readonly view: IViewStore;
    public readonly user: IUserStore;
    public readonly config: IConfigStore;
    public readonly router: RouterStore<IRootStore>;
    public readonly reports: IReportStore;
    public readonly dashboard: IDashboardStore;
    public readonly projects: ProjectStore;
    public readonly favorites: FavoriteStore;
    public readonly divisions: DivisionStore;

    constructor({
        auth,
        firestore,
        httpsCallable,
    }: {
        firestore: firebase.firestore.Firestore,
        httpsCallable?: (name: string) => firebase.functions.HttpsCallable,
        auth?: firebase.auth.Auth,
    }) {
        this.user = new UserStore(
            this,
            {
                auth,
                firestore,
            }
        );

        this.view = new ViewStore();
        this.config = new ConfigStore(
            this,
            {
                firestore,
            },
        );
        this.timesheets = new RegistrationStore(
            this,
            {
                firestore,
            },
        );
        this.reports = new ReportStore(
            this,
            {
                firestore,
            },
        );
        this.dashboard = new DashboardStore(
            this,
            {
                firestore,
            },
        );
        this.projects = new ProjectStore(
            this,
            {
                firestore,
            },
        );
        this.favorites = new FavoriteStore(
            this,
            {
                firestore,
            },
        );
        this.divisions = new DivisionStore(
            this,
            {
                firestore,
                httpsCallable,
            },
        );
        this.router = new RouterStore<IRootStore>(this);
    }

    public dispose() {
        this.config.dispose();
        // this.dashboard.dispose();
        // this.favorites.dispose();
        // this.projects.dispose();
        // this.reports.dispose();
        this.timesheets.dispose();
        this.user.dispose();
        // this.view.dispose();
    }
};
