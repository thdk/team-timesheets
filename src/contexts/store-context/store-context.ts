import { createContext } from 'react';
import { IRootStore } from "../../stores/root-store";

export const StoreContext = createContext<IRootStore>({} as IRootStore);
export const StoreProvider = StoreContext.Provider;
