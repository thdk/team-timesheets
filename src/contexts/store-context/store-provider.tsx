import React from "react";
import { useFirebase } from "../firebase-context/firebase-context";
import { StoreContext } from "./store-context";
import { Store, IRootStore } from "../../stores/root-store";
import { getAuth } from "firebase/auth";
import { getFirestore } from "firebase/firestore";
import { getFunctions, httpsCallable } from "firebase/functions";



export const StoreProvider = ({ children, testStore }: React.PropsWithChildren<{ testStore?: IRootStore }>) => {

    const firebase = useFirebase();

    if (testStore) {
        return (
            <StoreContext.Provider value={testStore}>
                {children}
            </StoreContext.Provider>
        );
    }

    const auth = getAuth(firebase);
    const firestore = getFirestore(firebase);
    const functions = getFunctions(firebase);

    const store = (window as any)["store"] = testStore || new Store({
        auth,
        firestore,
        httpsCallable: (name) => httpsCallable(functions, name),
    });

    return (
        <StoreContext.Provider value={store}>
            {children}
        </StoreContext.Provider>
    );
}
