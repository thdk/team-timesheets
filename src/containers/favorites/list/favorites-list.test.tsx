import React from "react";
import type firebase from "firebase";

import path from "path";
import fs from "fs";

import { FavoritesList } from ".";
import { StoreContext } from "../../../contexts/store-context";
import { Store } from "../../../stores/root-store";
import { render, waitFor } from "@testing-library/react";
import { loadFirestoreRules, clearFirestoreData, initializeTestApp } from "@firebase/rules-unit-testing";

const projectId = "favorites-list-test";
const app = initializeTestApp({
    projectId,
});

let store: Store;

beforeAll(async () => {
    await loadFirestoreRules({
        projectId,
        rules: fs.readFileSync(path.resolve(__dirname, "../../../../firestore.rules.test"), "utf8"),
    })
});

beforeEach(() => {
    store = new Store({
        firestore: app.firestore(),
    });
});

afterEach(async () => {
    store.dispose();
    await clearFirestoreData({
        projectId,
    });
})

afterAll(() => app.delete());

describe("FavoritesList", () => {

    it("should render without favorites", () => {

        const { asFragment, unmount, } = render(
            <StoreContext.Provider value={store}>
                <FavoritesList />
            </StoreContext.Provider>
        );

        expect(asFragment()).toMatchSnapshot();

        unmount();
    });

    it("should render favorites", async () => {
        const setupAsync = () => {

            return Promise.all([
                // favorites
                store.favorites.favoriteCollection.addAsync([
                    {
                        groupId: "group-1",
                        userId: "user-1",
                        description: "Favorite desc 1",
                        client: "client-1",
                        time: 3,
                    },
                    {
                        groupId: "group-1",
                        userId: "user-1",
                        description: "Favorite desc 2",
                        task: "task-1",
                        time: 2,
                    },
                    {
                        groupId: "group-1",
                        userId: "user-1",
                        description: "Favorite desc 3",
                        task: "task-1",
                        project: "project-1",
                        time: 2,
                    },
                ]),
                store.user.usersCollection.addAsync(
                    {
                        name: "user 1",
                        uid: "user-1",
                        email: "email@email.com",
                        roles: {
                            user: true,
                        },
                        divisionId: "",
                        tasks: new Map(),
                        recentProjects: [],
                    },
                    "user-1",
                ),
                store.config.clientsCollection.addAsync(
                    {
                        name: "client 1",
                        divisionId: "",
                    },
                    "client-1",
                ),
                store.favorites.addDocument(
                    {
                        name: "group 1",
                        userId: "user-1",
                    },
                    "group-1",
                ),
                store.projects.addDocument(
                    {
                        createdBy: "user-1",
                        name: "project 1",
                        divisionId: "",
                    },
                    "project-1",
                ),
                store.config.tasksCollection.addAsync(
                    {
                        name: "task 1",
                        divisionId: "",
                    },
                    "task-1",
                ),
            ]);
        };

        await setupAsync();

        store.auth.setUser({
            uid: "user-1",
            displayName: "user 1",
            email: "email@email.com",
        } as firebase.User);

        store.favorites.setActiveDocumentId("group-1");

        const { getByText, asFragment, unmount, } = render(
            <StoreContext.Provider value={store}>
                <FavoritesList />
            </StoreContext.Provider>
        );

        // FavoritesList will fetch favorites by groupId once it get's rendered
        // so we have to wait here for those favorites to be fetched and rendered
        await waitFor(
            () => expect(getByText("Favorite desc 2")),
        );

        await waitFor(() => expect(store.config.clientsCollection.isFetched).toBeTruthy());
        await waitFor(() => expect(asFragment()).toMatchSnapshot());

        unmount();
    });
});