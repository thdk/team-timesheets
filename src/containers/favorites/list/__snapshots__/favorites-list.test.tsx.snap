// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoritesList should render favorites 1`] = `
<DocumentFragment>
  <div
    class="data-row registration-line registration-line--readonly"
  >
    <div
      class="data-row__header"
    />
    <div
      class="data-row__column registration-line__time"
    >
      3.00
    </div>
    <div
      class="data-row__column registration-line__content"
    >
      <span
        class="data-row__line2 mdc-typography--subtitle2"
      >
        client 1 - Favorite desc 1
      </span>
    </div>
  </div>
  <div
    class="data-row registration-line registration-line--readonly"
  >
    <div
      class="data-row__header"
    />
    <div
      class="data-row__column registration-line__time"
    >
      2.00
    </div>
    <div
      class="data-row__column registration-line__content"
    >
      <span
        class="data-row__line2 mdc-typography--subtitle2"
      >
        Favorite desc 2
      </span>
    </div>
  </div>
  <div
    class="data-row registration-line registration-line--readonly"
  >
    <div
      class="data-row__header"
    />
    <div
      class="data-row__column registration-line__time"
    >
      2.00
    </div>
    <div
      class="data-row__column registration-line__content"
    >
      <span
        class="data-row__line1 mdc-typography--subtitle1"
      >
        project 1
      </span>
      <span
        class="data-row__line2 mdc-typography--subtitle2"
      >
        Favorite desc 3
      </span>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FavoritesList should render without favorites 1`] = `<DocumentFragment />`;
